torch/csrc/jit/codegen/cuda/lower_index.cpp

```diff
@@ -921,6 +921,103 @@ std::vector<kir::Allocate*> IndexLowering::allocateWelfordWorkBuffer(
   return work_buffers;
 }
 
+namespace {
+
+// Returns true if a GroupedWelfordOp op is eligible for using the
+// outer-optimized grouped welford runtime function
+bool canUseOuterOptRuntimeKernel(const GroupedWelfordOp* grouped_wop) {
+  const auto out_tv = ir_utils::getTvOutput(grouped_wop);
+  const auto out_domain = out_tv->domain();
+
+  if (!out_domain->hasGridReduction()) {
+    return false;
+  }
+
+  // TIDx and BIDx must be used for non-reduction domains. TIDy and
+  // BIDy must be used for reduction domains.
+  ParallelTypeBitmap used_pts;
+  for (auto leaf_id : out_domain->domain()) {
+    auto pt = leaf_id->getParallelType();
+    if (isParallelTypeThread(pt)) {
+      used_pts.set(pt);
+      if ((leaf_id->isReduction() &&
+           (pt == ParallelType::BIDy || pt == ParallelType::TIDy)) ||
+          (leaf_id->getIterType() == IterType::Iteration &&
+           (pt == ParallelType::BIDx || pt == ParallelType::TIDx))) {
+        // valid pattern
+        continue;
+      } else {
+        return false;
+      }
+    }
+  }
+
+  ParallelTypeBitmap valid_pt_map;
+  valid_pt_map.set(ParallelType::BIDx);
+  valid_pt_map.set(ParallelType::BIDy);
+  valid_pt_map.set(ParallelType::TIDx);
+  valid_pt_map.set(ParallelType::TIDy);
```

Comment:
Do you need to validate `BIDz` and `TIDz` are unused in the entire kernel?

Reply:
Fixed